src/Groups/GAPGroups.jl

```diff
@@ -0,0 +1,617 @@
+   else
+      return false, nothing
+   end
+end
+
+# END subgroups conjugation
+
+
+################################################################################
+#
+# Normal Structure
+#
+################################################################################
+
+"""
+```
+    normalizer(G::Group, H::Group)
+    normaliser(G::Group, H::Group)
+```
+returns `N,f`, where `N` is the normalizer of `H` in `G` and `f` is the embedding morphism of `N` into `G`.
+"""
+normalizer(G::T, H::T) where T<:Group = _as_subgroup(GAP.Globals.Normalizer(G.X,H.X),G)
+
+"""
+```
+    normalizer(G::Group, x::GAPGroupElem)
+    normaliser(G::Group, x::GAPGroupElem)
+```
+returns `N,f`, where `N` is the normalizer of <`x`> in `G` and `f` is the embedding morphism of `N` into `G`.
+"""
+normalizer(G::Group, x::GAPGroupElem) = _as_subgroup(GAP.Globals.Normalizer(G.X,x.X),G)
+
+normaliser = normalizer
+
+"""
+    core(G::Group, H::Group)
+returns `C,f`, where `C` is the normal core of `H` in `G`, and `f` is the embedding morphism of `C` into `G`.
+"""
+core(G::T, H::T) where T<:Group = _as_subgroup(GAP.Globals.Core(G.X,H.X),G)
+
+# normal_closure(G::T, H::T) where T<:Group = T(GAP.Globals.NormalClosure(G.X,H.X))
+# we don't know how G,H embeds into N_C(G,H) 
+
+"""
+    pcore(G::Group, p::Int64)
+return `C,f`, where `C` is the `p`-core (i.e. the largest normal `p`-subgroup) of `G` and `f` is the embedding morphism of `C` into `G`.
+"""
+function pcore(G::Group, p::Int64)
+   if !GAP.Globals.IsPrime(p)
+      throw(ArgumentError("p is not a prime"))
+   end
+   return _as_subgroup(GAP.Globals.PCore(G.X,p),G)
+end
+
+
+
+################################################################################
+#
+# Specific Subgroups
+#
+################################################################################
+
+# commutator_subgroup(G::T, H::T) where T<:Group = T(GAP.Globals.CommutatorSubgroup(G.X,H.X))
+# we don't know how G,H embed into [G,H]
+
+fitting_subgroup(G::Group) = _as_subgroup(GAP.Globals.FittingSubgroup(G.X),G)
+
+frattini_subgroup(G::Group) = _as_subgroup(GAP.Globals.FrattiniSubgroup(G.X),G)
+
+radical_subgroup(G::Group) = _as_subgroup(GAP.Globals.RadicalGroup(G.X),G)
+
+socle(G::Group) = _as_subgroup(GAP.Globals.Socle(G.X),G)
+
+
+################################################################################
+#
+# Sylow & Hall Subgroups
+#
+################################################################################
+
+"""
+    sylow_subgroup(G::Group, p::Int64)
+return a Sylow `p`-subgroup of `G`.
+"""
+function sylow_subgroup(G::Group, p::Int64)
+   if !GAP.Globals.IsPrime(p)
+      throw(ArgumentError("p is not a prime"))
+   end
+   return _as_subgroup(GAP.Globals.SylowSubgroup(G.X,p),G)
+end
+
+"""
+    hall_subgroup(G::Group, P::Array{Int64})
+return a Hall `P`-subgroup of `G`. It works only if `G` is solvable.
+"""
+function hall_subgroup(G::Group, P::Array{Int64})
```

Comment:
AbstractVector, Integer?

Reply:
If I put `Integer` instead of `Int64` or `Int`, it returns an error (`MethodError: no method matching` etc.)

Reply:
because unless You insist (writing explicitely `Integer[1,2,3]`) no array will be of type `Array{Integer}`. `Integer` is an abstract type and as such julia will try its best to avoid creating it when it can.

What you need to do is to make `P::Array{<:Integer}` which will match any array where elements are of type which is a subtype of `Integer`.

What would be even better is to make `P::AbstractVector{<:Integer}` where it would match any vector-like container with integer-like entries, e.g. `Int16(1):Int16(5)` is such a thing (`UnitRange`), but `big(1):2:5` (`StepRange`) is different. If your function depends only on indexing, or collecting and other vector-like behaviors, it's good to express it in widest possible signature to avoid future changes and duplicating code, etc.

Reply:
In the definitions of `perm` and `cperm` I replaced `Array{Int64}` by `AbstractVector{<:Integer}`. It still crashes with the same error message (`Method error: no method matching...`), except the case `function perm(L::AbstractVector{<:Integer})`: this works perfectly.

> julia> perm([2,3,1])
(1,2,3)

julia> perm(G,[2,3,1])
ERROR: MethodError: no method matching perm(::PermGroup, ::Array{Int64,1})

I can't explain why this happens.

Reply:
without showing us the code it's hard to judge ;)

most probably `perm([2,3,1])` works, since it calls `AbstractAlgebra` function which is exported to Oscar :P

Reply:
Yes. If `perm` is defined as `function perm(L::AbstractVector{<:Integer})`, then `perm([2,3,1])` belongs to an object of type `AbstractAlgebra.Generic.SymmetricGroup{Int64}`, instead of type `PermGroup` (the Oscar type). It seems that Julia does not recognize `Int64` as a subtype of `Integer`. Is this due to the redefinition of `Integer` in the file src/OscarTypes.jl ?

Reply:
I doubt it. Oscar.Integer only exists inside the Oscar module itself, and even then only for documentation purposes. It shouldn't affect GapGroups in any way. It doesn't actually redefine Base.Integer.

You can check if it is the problem by using Base.Integer instead of just Integer. But I'd be pretty surprised if that were causing any problems.

Reply:
Oh I see, GapGroups.jl is in Oscar.jl. I didn't realise that. Then yes, if you use Base.Integer, your problem will probably go away.

I will have to remind myself about how this was supposed to work. But for now, best to assume Oscar.Integer and Base.Integer are intended to be different things. Same for Rational, I think. But the craziness should end there, I think.

Reply:
Problem solved. With `Base.Integer` it works.

Reply:
the question is now if You want to shadow the defintion of `perm` for `AbstractAlgebra`

Reply:
Does it mean that we are no longer using the `AbstractAlgebra` objects for permutations?